k2/csrc/thread_pool.cu

```diff
@@ -0,0 +1,94 @@
+/**
+ * @brief Thread pool.
+ *
+ *
+ * @copyright
+ * Copyright (c)  2020  Xiaomi Corporation (authors: Fangjun Kuang)
+ *
+ * @copyright
+ * See LICENSE for clarification regarding multiple authors
+ */
+
+#include <utility>
+
+#include "k2/csrc/thread_pool.h"
+
+namespace k2 {
+
+static int32_t GetDefaultNumThreads() {
+  int num_threads = std::thread::hardware_concurrency();
+  if (num_threads == 0) num_threads = 1;
+  return num_threads;
+}
+
+ThreadPool::ThreadPool(int32_t num_threads)
+    : threads_(num_threads > 0 ? num_threads : GetDefaultNumThreads()) {
+  for (auto &thread : threads_)
+    thread = std::thread([this]() { this->ProcessTasks(); });
+}
+
+ThreadPool::~ThreadPool() {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    keep_running_ = false;
+    not_empty_cond_.notify_all();  // wake up all threads in the pool
+  }
+  for (auto &thread : threads_) thread.join();
+}
+
+void ThreadPool::WaitAllTasksFinished() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!finished_) {
+    // wait for the `empty_cond_` condition.
+    empty_cond_.wait(lock);
+  }
+}
+
+void ThreadPool::ProcessTasks() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (keep_running_) {
+    while (tasks_.empty() && keep_running_) {
+      // Wait for the `not_empty_` condition.
+      // RunTask() or the destructor will signal it.
+      not_empty_cond_.wait(lock);
+    }
+
+    if (!keep_running_) break;
+
+    {
+      auto task = std::move(tasks_.front());
+      tasks_.pop();
+      if (!tasks_.empty()) not_empty_cond_.notify_one();
```

Comment:
I think we don't need to notify `not_empty` here, `RunTask` should do this and it has done this.

Reply:
In which case do other threads have a chance to run in case the queue is not empty, but  `RunTask` is not called and this statement is removed?

Reply:
mm, sorry, I didn't understand your question here, the point here is: thread pool is a typical producer-consumer model, so I think it's very clear that `SubmitTask` just push task and notify, the `ProcessTask` just consume the task. I don't what's your point here to `notify` in a consumer, i.e. ProcessTask

Reply:
>  I think it's very clear that SubmitTask just push task and notify, the ProcessTask just consume the task.

Can the consumer always be notified? Is it possible that the notification gets lost?

If all consumers are running, i.e., no consumer is in waiting state, does the `notify` event from the producer
have any effect?